File: backend/app/api/v1/routes/payments.py
"""
Endpoints de pagos (Mercado Pago - Card Payment Brick).

`POST /v1/payments/create` recibe el `cardFormData` que el Brick de
MercadoPago emite desde el navegador del cliente (con el `token`
opaco) y crea un `Payment` real en la API de MP. El backend NO
recibe los datos sensibles de la tarjeta (PAN, CVV, vencimiento):
el Brick los tokeniza en su iframe.

Si el pago es aprobado, el frontend debe llamar inmediatamente a
`/v1/bookings/process` enviando `mp_payment_id` para que la reserva
quede vinculada al pago real.
"""

import logging

from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.orm import Session

from app.core.database import get_db
from app.schemas.payment_schema import (
    CardPaymentCreateRequest,
    CardPaymentCreateResponse,
)
from app.services.payment_service import MercadoPagoError, PaymentService

logger = logging.getLogger(__name__)

router = APIRouter()


@router.post(
    "/create",
    response_model=CardPaymentCreateResponse,
    status_code=status.HTTP_200_OK,
    summary="Crear pago con tarjeta (Card Payment Brick - sandbox)",
    tags=["Payments"],
)
async def create_card_payment(
    payload: CardPaymentCreateRequest,
    db: Session = Depends(get_db),
) -> CardPaymentCreateResponse:
    """
    Crea un `Payment` en Mercado Pago con el token del Brick.

    Devuelve el `id` y `status` del pago. Si `approved=true`, el
    frontend debe llamar a `/v1/bookings/process` con el campo
    `mp_payment_id` para emitir la reserva. Si `approved=false`, se
    muestra el `message` al usuario (sin emitir reserva).
    """
    try:
        service = PaymentService(db=db)
        return service.create_card_payment(payload)
    except MercadoPagoError as exc:
        # Errores controlados (credenciales faltantes, MP caído, etc.)
        logger.error("[MP] Error controlado: %s", exc)
        raise HTTPException(
            status_code=exc.status_code,
            detail=str(exc),
        ) from exc

File: frontend-client/src/api/payments.js
import axiosInstance from './axiosInstance'

function unwrapError(err) {
  if (err && typeof err === 'object' && 'message' in err) {
    return err
  }
  return new Error('Error inesperado al comunicarse con el servidor.')
}

/**
 * POST /v1/payments/create
 * Crea un `Payment` en Mercado Pago con el token que emitió el
 * Card Payment Brick. NO se envían datos sensibles de la tarjeta
 * al backend: el Brick ya los tokenizó en su iframe.
 *
 * Devuelve { payment, approved, message }.
 * - `approved=true`  → llamar a /v1/bookings/process con `mp_payment_id`.
 * - `approved=false` → mostrar `message` al usuario.
 */
export async function createCardPayment(cardFormData) {
  const cf = cardFormData || {}

  // FIX: el Brick puede enviar los campos en camelCase o snake_case
  // según la versión. Aceptamos ambos para ser defensivos.
  const token = cf.token
  const paymentMethodId =
    cf.paymentMethodId || cf.payment_method_id || cf.paymentMethod || cf.payment_method
  const issuerId = cf.issuerId || cf.issuer_id
  const installments = cf.installments ?? 1
  const transactionAmount =
    cf.transactionAmount ?? cf.transaction_amount ?? cf.amount
  const externalReference =
    cf.externalReference || cf.external_reference || cf.external_ref
  const payer = cf.payer || {}
  const cardholderName = cf.cardholderName || cf.cardholder_name

  const payload = {
    token,
    payment_method_id: paymentMethodId,
    issuer_id: issuerId || undefined,
    installments: Number(installments) || 1,
    transaction_amount: Number(transactionAmount),
    external_reference: externalReference || 'BOOKING:NA:NA:1',
    payer: {
      email: payer?.email,
      first_name: cardholderName || payer?.first_name || undefined,
      identification_type: payer?.identification?.type || undefined,
      identification_number: payer?.identification?.number || undefined,
    },
  }

  try {
    const { data } = await axiosInstance.post('/payments/create', payload)
    return {
      payment: {
        id: data?.payment?.id,
        status: data?.payment?.status,
        statusDetail: data?.payment?.status_detail,
        paymentMethodId: data?.payment?.payment_method_id,
        externalReference: data?.payment?.external_reference,
      },
      approved: Boolean(data?.approved),
      message: data?.message || '',
    }
  } catch (err) {
    throw unwrapError(err)
  }
}
